Fix findLabel() performance being bad if labels are missing and being O(n^n) when they don't have tag id like ID-lbl

ggppdk opened the pull request:
#### Summary of Changes

If you have a lot of form elements e.g. buttons that do not have a label or if the form has labels that do not have HTML tag ID like *-lbl, then findLabel() performance does a full DOM search using selector:
var $label = $('label[for="'+$elem.attr('id')+'"]');

The above describe findLabel() behaviour of:
- validate.js

For:
- html5fallback.js
  the full DOM search is always performed
#### Testing Instructions

Apply the patch the edit file (or some other form file):
and replace:

```
</form>
```

with:

```
<?php
echo '
<script>
    var test_cnt = 0;

    function time_test_Isvalid(el)
    {
        window.console.log( \'CALLING isValid()\' );

        var vTimeStart = new Date();
        document.formvalidator.isValid(document.getElementById(el.form.id));
        var vTimeDiff = (new Date())  - vTimeStart;

        window.console.log( \'isValid() time: \' + vTimeDiff );

        return false;
    }

    function inject_test_input(el, box_id)
    {
        var box = document.getElementById(box_id);

        var lbl = document.createElement(\'label\');
        lbl.setAttribute(\'for\', \'testid_\' + test_cnt);
        lbl.setAttribute(\'class\', \'label\');
        lbl.innerHTML = \'New element\';
        box.appendChild(lbl);

        var input = document.createElement(\'input\');
        input.type = \'text\';
        input.setAttribute(\'class\', \'required\');
        input.id = \'testid_\' + test_cnt;
        box.appendChild(input);

        test_cnt++;
        window.console.log(\'Element injected\');
        return false;
    }   
</script>

<br/>
<button  onclick="time_test_Isvalid(this);  return false;"  class="btn">  Test validation </button>
<button  onclick="inject_test_input(this, \'inject_test_box\'); time_test_Isvalid(this); return false;"  class="btn">   Inject field + Test validation </button>
<div id="inject_test_box"></div>
';
$use_labels = 1;
$add_for = 1;
$add_dash_lbl = 1;
for($i = 0; $i < 10000; $i++)
{
    if ($use_labels)
    {
        $lbl_for = $add_for  ?  'for="dummy_field_'.$i.'"'  :  '';
        $lbl_id  = $add_dash_lbl  ?  'id="dummy_field_'.$i.'-lbl"'  :  'id="dummy_field_'.$i.'-nodashlbl"';
        echo '<label '.$lbl_for.' '.$lbl_id.' class="label">Element '.$i.'</label> ';
    }
    echo '<input id="dummy_field_'.$i.'" type="text" name="dummy_field_'.$i.'" /> ';
}
?>

</form>
```

The above will 
- add 10,000 form elements into the form
- add a button to trigger and TIME validation of isValid() (validate.js) without submitting the form
  (open console to see output)
- add a button to inject a new form element into the DOM and re-trigger validation, so that we can verify that validation behaviour and performance after validation has been trigger once and some element has been added after it

**Validations** (testing times incude only isValid() of validate.js):

```
All fields have labels with ID-lbl
             1st      2nd, +           After Injecting new field
BEFORE:      1070ms    about 250ms      250ms - 550ms
AFTER:        630ms    about 250ms      370ms - 630ms

All 10,000 labels are missing, e.g. input buttons or buttons
             1st      2nd, +           After Injecting new field
BEFORE:     5700ms   about 180ms    180ms - 430ms
AFTER:       450ms   about 180ms    230ms - 480ms


All 10,000 labels don't have id like ID-lbl
             1st       2nd, +         After Injecting new field
BEFORE:   28700ms    about 220ms      220ms - 440ms
AFTER:      620ms    about 220ms      360ms - 650ms
```

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+		// Iteration through DOM labels
+		var $lbl, $el, for_id, el_id;
+		var $lbls_hash = {};
+		jQuery('label').each(function()
```
I would suggest to restrict the search by current form:
```javascript
jQuery(form).find('label')
```

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+				el_id = $el.attr('id');
+				(el_id && $lbls_hash.hasOwnProperty(el_id)) ?
+					$el.data('label', $lbls_hash[el_id]) :
+					$el.data('label', $empty) ;
```
I would not play so tricky :wink: 
just set it to `false`, later it will be more easily to test whether the label exists or not by:
```javascript
$el.data('label') === false;
```
instead slower (and not safe btw):
```javascript
$el.data('label').length !== 0;
```

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+  	
+   	// New element encountered (first run or / newly injected into the dom), redo iteration of DOM labels ... updating this and any other injected elements
+  	else
+  		refreshFormLabels(form, form.elements);
```
you do not need `form.elements` as `form` is already array of elements, and you can iterate though `form`:
```javascript
for(var i = 0, l = form.length; i < l; i++){
  console.log(form[i]); // will show form input
}
```

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+		});
+
+		// Set to zero length the .data('label') of elements without one
+		if (typeof f === 'undefined') f = form.elements;
```
look my comment about `form.elements` :wink:

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+			var $parentElem = $elem.parent();
+		  parentTagName = $parentElem.get(0).tagName.toLowerCase();
+		  
+			if(parentTagName == "label")
```
do not need `toLowerCase()`, just compare with `tagName === 'LABEL'` :wink:

Fedik commented:
looks like not bad idea :wink:

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+		if ( !!$label ) ;  // $label && $label !== undefined
+		
+		else if (!id)
+			$elem.data('label', jQuery());
```
here also, just use `false`

ggppdk commented:
@Fedik, thanks for your comments ,

About
> jQuery(form).find('label')

Actually if i had thought of it and **i wish it could be like that** , but it can not be ...
According to HTML specifications ... fields and their label can be placed outside:
<form> ... </form>
- and the way it is done is by using the form attribute  
```
<input ...  form="form-TAG-ID" />
```
... which gives me the opportunity to say that current isValid() implementation needs fixing this !

About not using:
$el.data('label') === false;
- used it like this === undefined to be backwards compatible ..., keep existing behaviour and avoid breaking some 3rd party code that might be overriding only 1 function of the validator ? e.g. isValid ?

about 
parentTagName = $parentElem.get(0).tagName.toLowerCase();
- we already use it in other places
- it computation effect is minimal and used only for some fields only
- not all document types guarantee it to be uppercase
- in future the standards maybe different

about no need to use form.elements i am not sure, i will check, 
- i remember needing this to properly work with <fieldsets class="radio">   and class="checkbox"

but i will check, i am not sure if it is needed, but anyway it will not provide any speed benefit (i think)

Fedik commented on media/system/js/validate-uncompressed.js:
```diff
@@ -22,28 +22,81 @@ var JFormValidator = function() {
+ 	{
+		if (!id) return false;
+		
+ 	 	var $elem = jQuery('#'+id);
```
hm, it does not looks good, because we already have the element before call `findLabel()`, why we need to search it again.
Maybe would be good idea change the call from `findLabel($el.attr('id'), $el.get(0).form);` to `findLabel($el, $el.get(0).form);`

for keep old behavior, we could try do check, something like:
 ```
if id is Element:
   then use it
else if id is string:
   then search element by id
```

Fedik commented:
> According to HTML specifications ... fields and their labels can be placed outside

maybe it could be, but it never happens in Joomla world, I think 

> $el.data('label') === false;
> keep existing behaviour and avoid breaking some 3rd party code

I do not see b/c problem here.
Before use `$el.data('label').length` they already should check that `$el.data('label')` exists, other way the code `$el.data('label').length !== 0` will fail, when in some contingency the label will not exist (can be removed by other js or so)
JavaScript is dynamic thing :wink: 

> parentTagName = $parentElem.get(0).tagName.toLowerCase();
> not all document types guarantee it to be uppercase

one more note, you need to use `nodeName` instead `tagName`, then all be fine

ggppdk commented:
-1- about ... searching label only inside < form > using:  jQuery(form).find('label')
- it is better to be compliant (if / when possible, also little performance gain to break standards)
- but also current Joomla code checks all the DOM, so ... we would change old behaviour ))

so better let it be compliant

-2- about using nodeName instead of tagName, i agree with you that it is better
- just for the purpose of checking if the node is a LABEL , **it makes no difference**
in other case yes it would make difference e.g. if checking if element is a **text node**
- nodeName also supports these nodes: **attribute nodes, text nodes, comment nodes**

i will replace it (still current code works)

-3- 
about checking
$label =  $elem.data('label');
and checking .length
-  it was like this before  ... FindLabel (current joomla code)  always return a jQuery object

**so we do not need to check if it is null / false / undefined**
because the new code of findLabel() does the same thing as the previous
- it will return a jQuery object ... and if label was not found then it returns zero length one

ggppdk commented:
@Fedik 

I just saw that i responed to a previous message in a confusing way
i am sorry, i wrongly said:
> ... $elem.data('label')   ....
> - === false means field does not have a label

**i meant if findLabel() returns false then label was not found**

$elem.data('label') can only be these
- undefined  (label not yet searched)
- jQuery Object  with non-zero length  (aka label was found)
- jQuery Object  with zero length  (aka label was NOT found)

now 2 things:

-1- about your suggestion to setting $elem.data('label')  to **false instead instead of empty jQuery object**
- there is other code that examines it like in validateForm() of html5fallback.js, and maybe 3rd party code ? (i don't know)

so it is like the above for B/C, thus we much match the behaviour of what the selector would do:
$form.find('label[for="' + id + '"]');
it would return a zero length jQuery object if label was not found


-2- but your suggestion to check for non-null / not undefined value after refreshFormLabels() makes sense in ... case that refreshFormLabels() is bogus, or in case it was mis-used,
- i will add check for it too,

but still it should never happen with current code

Fedik commented:
well,  maybe you are right about use empty jQuery object in `$elem.data('label')`, 
but still, before check `$elem.data('label').length` there always need to check that `$elem.data('label')` actually returns some value, cause the label easily can be removed by external script

ggppdk commented:
yes, i will add check for it

about removing label by external script and letting the field in place
you are right

- but it is usually the opposite, leave the label in place and replace the field, OR ... replace both, in these cases the code will work
- **plus current Joomla JS code does not check if the label has been removed** !, current Joomla code will use whatever is stored inside .data('label')
- and here we are discussing about if .data('label') is set to some value (removed by 3rd party script ??), and not discussing if the label has been removed, which i thought to add check, but since current Joomla code does not check and because it would code extra, i have not added it

now, i am adding both of your suggestions 
- for checking if .data('label') is non-empty
- replacing tagName for nodeName

updating html5fallback.js too

ggppdk commented:
@Fedik 

i was about to replace tagName with nodeName, and i have just thought that the code may not be needed at all !
(and removing it, would be good for performance too, in case of many label missing e.g. buttons)

```
if ($label.length <= 0)
{
	var $parentElem = $elem.parent();
  parentTagName = $parentElem.get(0).tagName.toLowerCase();
  
	if(parentTagName == "label")
	{
		$label = $parentElem;
		$elem.data('label', $label);
	}
}
```

look at it ... it catches the case that you are containing a **input inside a label** TAG which does not have either of atributes:   id="ID-lbl" and for="", otherwise it would have been found already
- i added because it exists in html5fallback.js

so because
- there is no such joomla code to create such a thing  (there is no such form element)
- and **because current JS code of validate.js is not checking it**

i think to remove it, not encourage someone writting and using this case and also saving improving performance in some cases

but it does exist in existing code of html5fallback.js,  shall i leave it for html5fallback.js  ?

ggppdk commented:
@Fedik 

i have changed the
elem.data('label') to store **false** if label is missing **as you had suggested**
- that is how it was before this pull request, but i was trying to make the elem.data('label') to be shared with the findLabel() of html5fallback.js, because for html5fallback.js the **return value of findLabel()** is an empty jQuery object instead of false,

To summarize new changes / fixes
1. now it is doing both:
-- setting elem.data('label') to **false** if label is missing 
-- elem.data('label')  is now properly shared with html5fallback.js, in case that html5fallback.js is re-used in the future (your recently merged PR has suppressed its validateForm())
2. also added a check for the case that an invalid field has no label, current behaviour is not add any message at the top of the page, with this new check if label is missing the invalid field message will be added inline
3. also removed the case of checking **input inside label that lucks both id="ID-lbl" and for="ID"**, no Joomla code is using this case, i had added it because it existed in html5fallback.js

Fedik commented:
@ggppdk I not sure whether worth to do anything with html5fallback.js
but thanks for your work, I try test more, as soon as find a time

ggppdk commented:
Hello

fixed the issue with form.elements not being used properly,

i found when bug was introduced,
- it was when mootools code of validation.js was replaced with jQuery code 

https://github.com/joomla/joomla-cms/commit/763c69f4bedc38f62ddf4f04af9d3d7fe3f3a719

Fedik commented:
@ggppdk good find :+1: 
also I see that the old code have searched the labels in similar way as you have suggested :wink: 
```javascript
if (!(el.labelref)) {
    var labels = $$('label');
    labels.each(function(label){
        if (label.get('for') == el.get('id')) {
            el.labelref = label;
        }
    });
}
```
but without caching

ggppdk commented:
About old code being similar
- yes similar,  in terms of both using a label iterration, but that is where the similarity stops
- the main difference is that my code (both) **does and guarantees** label iteration to only happens **once**, 

#### Because it 
1. during iteration it updates all form fields (and the fieldset class=radio/checkbox) and not only a single field
2. and also for fields without a label, a "false" value is set to indicate no label found and **thus avoiding the iteration to be triggered again**

- Iteration will never be triggerred again unless the DOM is modified to inject a new form field

That old code was slow
so the existing Joomla code
- replaced it with  **HTML tag id look-up:  ID-lbl**

**which is fast** becaue of the TAG ID look,
but if a field does not have a label with (**ID-lbl**) 
- then the fallback is a full DOM search with the selector:
```
var $label = $('label[for="'+$elem.attr('id')+'"]');
```

Which this PR fixes

Fedik commented:
I have made some test, and a I am confused :smile:
The result not like I have expected.

I have made test in Global Configuration form.

### Test  the form with 200 of User groups (around 3000 inputs)
**Before patch**
first validation 143ms
![3n-1n](https://cloud.githubusercontent.com/assets/1568198/15367669/d0cc2646-1d32-11e6-8c18-a2a851cfde4b.png)
second validation 31ms
![3n-2n](https://cloud.githubusercontent.com/assets/1568198/15367708/ed222174-1d32-11e6-9b01-b84ddb353217.png)

**After patch**
first validation 164ms
![3y-1y](https://cloud.githubusercontent.com/assets/1568198/15367750/12f05952-1d33-11e6-804f-17d968b5b32b.png)
second validation 43ms
![3y-2y](https://cloud.githubusercontent.com/assets/1568198/15367760/1d6c4148-1d33-11e6-8bc3-43bed0056f30.png)

In result the suggested patch is "slower".

### Test the form with 200 of User groups + 10000 inputs (around 13000 inputs)
Here I have added the inputs as in the testing instruction.

**Before patch**
first validation 875ms
![4n1](https://cloud.githubusercontent.com/assets/1568198/15367840/7235d946-1d33-11e6-8087-9c473119ce0e.png)
second validation 296ms
![4n2](https://cloud.githubusercontent.com/assets/1568198/15367855/87ee932c-1d33-11e6-859c-1cbdf1b4aeb0.png)

**After patch**
first validation 1010ms
![4y1](https://cloud.githubusercontent.com/assets/1568198/15367882/9eeb7662-1d33-11e6-8a10-69a57bad1baa.png)
second validation 130ms
![4y2](https://cloud.githubusercontent.com/assets/1568198/15367906/bee1c4bc-1d33-11e6-90ae-cea209c97904.png)

In result the suggested patch have won only second "validation".

No idea :smile: 
Maybe I have missed something.

ggppdk commented:
Hello

the degraded in both cases is small and it is expected
(see explanation below)

- if you have a form that all labels have 
ID-lbl you will see either no benefit or small degrade  (as you see above)

- also if you have a form with many "excluded" fields (e.g. ACL for 200 usergroups) again you will see small degrade on 1st validation because , the loop to find loop does not exclude them and makes a single pass for them

in any case the degrade is 10-15%

but for cases that this PR describes

**the existing code can be up to 40x slower** which huge performance issue for existing code
and that is what this PR fixes

ggppdk commented:
Also at the time i forgot to say
- you see 10%-15% degrade , and i explained why above,

but i forgot to say why your test **is different than my initial posting in this PR**

your tests did not included validateForm()  from html5fallback , right ?

- at the time i tested the  validateForm() of html5fallback  was still enabled / running

still this PR fixes performance issue with these 2 cases
- many labels missing
- many labels do not have for="ID-lbl"

in all other cases you will see small performance degrade on 1st run, since the validateForm() of ... html5fallback no longer runs

ggppdk commented:
@Fedik 

Thanks for the time you took to test this

1 more comment:
every time the cases that this PR fixes,  are found by a user
- then you get an performance issue reported here,
- then a PR follows to fix the issue (e.g. exclude fields from the validation loop)

if we would rather do the above, or if we do not care of forms of 3rd party extensions being slow, because they have not "patched" their forms in a similar way, then we may ignore this PR

Fedik commented:
hm yes, without `validateForm()`, on latest staging branch 
but case when both `validateForm()` and `isValid()` executed is a :bug: , should be executed only one validation.

ggppdk commented:
Yes  i don't say to re-enable validateForm() ,

it only needs to run to do cleanups, 
- only cleanup that i have seen is: remove placeholder text in no-supported browsers (IE8)

but anyway placeholder="..." is broken in IE8 (will cause a JS error into html5fallback.js) 

so the text never gets added inside the input,

so indeed there is no need to remove it on submit either

ggppdk commented:
@Fedik 

you have only made 1 of my 3 tests, there are 2 more test, please try 10,000 inputs with them:


#### Labels missing
```
$use_labels = 0;
$add_for = 1;
$add_dash_lbl = 1;
```

#### Labels not having id="ID-lbl"
```
$use_labels = 1;
$add_for = 1;
$add_dash_lbl = 0;
```

Fedik commented:
Ok, I have made one more try :smile:
And in this synthetic test the result more visible

### 1 From test instruction
```
$use_labels = 1;
$add_for = 1;
$add_dash_lbl = 1;
```

**Before patch:**
first validation 844ms:
![1n1_](https://cloud.githubusercontent.com/assets/1568198/15449137/c066dae2-1f7d-11e6-9bc5-bc55eb6f36b5.png)
second validation 280ms:
![1n2_](https://cloud.githubusercontent.com/assets/1568198/15449139/c6b59d34-1f7d-11e6-9ec0-6f15402817d0.png)

**After patch:**
first validation 762ms:
![1y1](https://cloud.githubusercontent.com/assets/1568198/15449143/f991aa22-1f7d-11e6-957e-276f4bba2308.png)
second validation 328ms:
![1y2](https://cloud.githubusercontent.com/assets/1568198/15449145/fd3ac49c-1f7d-11e6-90cb-2fa59b305023.png)

### 2 Labels missing
```
$use_labels = 0;
$add_for = 1;
$add_dash_lbl = 1;
```
**Before patch:**
first validation 9s:
![2n1](https://cloud.githubusercontent.com/assets/1568198/15449163/a1f2fde2-1f7e-11e6-99fa-372991aa25b2.png)
second validation 245ms:
![2n2](https://cloud.githubusercontent.com/assets/1568198/15449164/a7ed2682-1f7e-11e6-9787-423f31bbab6e.png)

**After patch:**
first validation 569ms:
![2y1](https://cloud.githubusercontent.com/assets/1568198/15449165/adc40b48-1f7e-11e6-9255-80e621787034.png)
second validation 299ms:
![2y2](https://cloud.githubusercontent.com/assets/1568198/15449167/b2318264-1f7e-11e6-829b-d82420b6d543.png)


### 3 Labels not having id="ID-lbl"
```
$use_labels = 1;
$add_for = 1;
$add_dash_lbl = 0;
```

**Before patch:**
first validation 27s:
![3n1](https://cloud.githubusercontent.com/assets/1568198/15449168/be3111ec-1f7e-11e6-8730-69996f76db41.png)
second validation 288ms:
![3n2](https://cloud.githubusercontent.com/assets/1568198/15449169/c51c3e1e-1f7e-11e6-9e52-8dd58c0cc26b.png)

**After patch:**
first validation 771ms:
![3y1](https://cloud.githubusercontent.com/assets/1568198/15449171/cd5a89e6-1f7e-11e6-8269-7c2d0a6bdcc6.png)
second validation 341ms:
![3y2](https://cloud.githubusercontent.com/assets/1568198/15449172/d17b0758-1f7e-11e6-8ce8-abc0cf133dbe.png)

though it is synthetic, I guess this can be count as success test :smile:

Fedik commented:
I have tested this item :white_check_mark: successfully on cef7b5db489f09666ebd99bbe23b3b55dfee7e6e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10136">issues.joomla.org/joomla-cms/10136</a>.</sub>

ggppdk commented:
@Fedik 
Thanks for all your comments, i will re-do and split this PR into 3
- label search performance
- fix for not all HTML form fields including in validation
- better (JS) validation messaging 

i am already using these in custom software
